default.nix

```diff
@@ -197,21 +198,22 @@ let
     ];
 
   # Function that applies enabled option to the package, used in the fold.
-  funcOnSwitchAppliesFunction = set: object:
+  onSwitchApplyFunc = set: object:
     if set.switch
       then set.function object
       else object;
 
   # General description of package
   package = haskellPackages.developPackage {
-    name = "hnix";
-    root = pkgs.nix-gitignore.gitignoreSource [ ] ./.;
+    name = cabalName;
+    # Do not include into closure the files listed in .gitignore
+    root = packageRoot;
 
     overrides = self: super: {
       # 2020-06-26 Due to a behaviour change in neat-interpolation-0.4, we
       # require n-i >= 0.4. dontCheck helps us avoid conflicts with
       # neat-interpolation's test dependencies.
-      neat-interpolation = pkgs.haskell.lib.dontCheck super.neat-interpolation_0_5_1;
+      neat-interpolation = pkgs.haskell.lib.dontCheck super.neat-interpolation_0_5_1_1;
```

Comment:
Mustn't we assume that `neat-interpolation_0_5_1_1` will soon be gone too, when the next version of `neat-interpolation` is released or the latest version becomes the default?

Reply:
(That's why I had used `callPackage` in https://github.com/haskell-nix/hnix/pull/680)

Reply:
You start to understand, this is a current design flaw of the Nixpkgs.

If someone knows how we can circumvent this flaw and always find import the last release version of package - I would be grad to know also.

Reply:
Argument is to prefer Nixpkgs packages over custom overrides, because we test Nixpkgs builds, and if we override everything - that is not Nixpkgs test, that is custom Nix config test.

Also preferring override of the package that available in the store essentially means to prefer avoiding the global store for what it was created and prefer keeping custom store, then the question is why we even test for Nixpkgs then if we had chosen to avoid it.

Reply:
Well, on the other side - I also do not like the Nixpkgs quirks constantly break the CI pass through.

As I said - lets make Nixpkgs optional. - That seems the most effective solution to the problem.

And then think maybe compose some middle ground, we can have a build option with frozen Nixpkgs overrides, so if devs/users can not build with current Nixpkgs - README can direct them to use frozen version. Of course, we can just direct to other GHC sets - that solves a part of that probabilistic issue, and of course we would need to maintain the frozen overrides if we provide them.

So currently making more Nixpkgs build optional seems the most effective solution.

Reply:
If we can make the Nixpkgs build optional, we might as well get of it.

I thought the basic problem was simply that the Nixpkgs revision is unpinned, and therefore each build might use a different and often incompatible configuration. I've never understood why we don't simply pin the revision and update it every few months.

Reply:
But the meaning of Nixpkgs build is Nixpkgs build tracking. Otherwise - we can just `cabal2nix` into custom Nix file and forget about Nixpkgs at all.

The argument that optional build can be reduced, it is analog to saying: since monitoring is optional - we can remove monitoring, since documentation is optional for operations - we can remove documentation.

Reply:
The reason for Nixpkgs to be optional - is that since it is brittle - we would see the changes that needs to be made, but CI would pass PRs at the same time, and we can see that Cabal & Hackage & testing suites work and that error relates to downstream Nipkgs issue and not to the PR code, and so would remind us to fix it.

The idea is so we shift from sequence requirement `(code merge request with CI block due to Nixpgs -> PR on Nixpkgs fix -> rebase CI tests & code merge -> HNix)` develop and fix Nixpkgs at the same time in parallel PRs `(code merge request -> HNix),  (PR on Nixpkgs fix -> HNix)`

Reply:
As I said, using of Nix for Nixpkgs testing is essentially indeed optional, it is HNix being a good citizen for Nix ecosystem.

Reply:
Regarding why we do not pin the Nixpkgs - the same question is "Why we do not freeze the Cabal configuration for Hackage."

Because there is no sense - if we pin Nixpkgs for our builds, then HNix would not work in new Nixpkgs, and then what is the sense in pinning Nixpkgs if all community constantly rolls on it and HNix does not work for them?

HNix can pin Nixpkgs to achieve interpreting Nixpgs successfully - then we internally stabilize the state so we can develop for it, but then we do not track interpretation of current Nixpkgs and would need to adjust to it afterward constantly anyway.

1. So pinning for Nixpkgs to build HNix - results into us not supporting and not advertising project to Nixpkgs community which is where main community located. 2. And pinning Nixpkgs for interpretation in HNix - means project most probably does not work for current Nixpkgs and we can not open Reports upstream, since we do not use current Nixpkgs and some old frozen one, and we anyway would need to follow Nixpkgs always, since it is a technical target of the project.

Reply:
OK. I'm somewhat skeptical about the need to build with the latest Nixpkgs all the time.

My current understanding is that the Haskell packages in Nixpkgs are basically just downstream from Stackage. So it seems to me that we can simply try to keep Stackage happy, and let the Nixpkgs maintainers worry about the rest.

If you still want this monitoring, that's fine by me, but please ensure that it doesn't affect CI results and that new contributors can get a clean build with the Nix setup at all times.

Reply:
How Nixpkgs can be downstream from the Stackage, if Nixpkgs autoimports downstream from the Hackage.

For example, HNix is on Hacakge, and not in Stackage, but in Nixpkgs, how it gets there after we release it on Hackage, in the night it already appears in Nixpkgs.

`stack2nix` was working half way half a time, and it is recenly deprecated by creators, `stackage2nix` also abandoned project.

The `peti` build projects `hackage2nix` `cabal2nix`, Nixpkgs machinery uses them, it generate `configuration-hackage2nix.yaml` that is the entry point for auto imported Haskell Nixpkgs project descriptions. He has presentations on it, and indeed every 3:30 in the night autoimported Haskell packages in Nixpkgs sync to Hackage, that is why HNix updates in there and we see if it builds or there is breakage.

---

The thing is. Nix is always pain. FP OS management for many years ahead would be a pain, it would be pain until FP OS ecosystems mature (and it would be longer process then Haskell maturity, FP OS management is much harder ad-hock question and volume of particulars then the design of the programming language), it would be a pain until IT would not check to comply with it, I mean that their projects build in Nix or Guix. All hidden under the carpet troubles of IT would and collaboration of its people, while hidden in classical approaches - would lay explicit marks on the newborn sick FP OS management child.

And as I stated, FP IT management is a much more complex topit - too much projects, project options, build systems, languages, own language, own store, hardware particulars - there is too much questions and too many work to maintain all of it while FP OS management is so demanding - it would mature for a long time.

It is and would be a constant PITA for a very long time, years and years. What I hope is that with migration on Haskell things would become bearable and velocity would have bigger acceleration, so much more exponentially rapid process would happen.

Reply:
> For example, HNix is on Hacakge, and not in Stackage, but in Nixpkgs, how it gets into Nixpkgs after we release it on Hackage, in the night it already appears in Nixpkgs.

Good question. I don't know. See https://github.com/haskell-nix/hnix/issues/676#issue-656518706 though, specifically

The problem is, that we get an older neat-interpolation ==0.3.2.6 from stack.

Reply:
That means that I need to learn more about Haskell Nix stack.

There is indeed Stack hooks.

Maybe part of the answer is that Stackage essentially is a Hackage, at least in terms of API, so same Hackage tools can download from both places.

Or most probably it is even simpler. Tool to get the consistent package set - receives a last LTS package version numbers from Stackage, and integrates those packages directly from Hackage or Stackage.

The full chain of `Hackage -> ... -> Nixpkgs install` is something I personally never thoroughly went through.

And actually, that is what I want to do, in previous talks I was talking to have a real Nixpkgs integration test, where the chain of integration woud be a great representation of Hackage -> Nixpkgs integration build and test process, it would revolve around `hackage2nix`, `cabal2nix` and regeneration of `configuration-hackage2nix.yml`